### PR TITLE
Feature/unit tests

### PR DESCRIPTION
**Pull request by dneukirchen**
This PR brings the unit tests back to the core as discussed in https://github.com/joomla/joomla-cms/pull/23962

* added psr-4 autoloading for tests using composer 
* namespaced unit tests
* added unit tests 
* refactored existing unit tests
* removed useless unit tests
* re-integrated some of the unit tests that are currently not running
* added UnitTestCase as a base class
* updated build tools
* fixed one bug in Mail class which surprisingly was not detected by previous unit tests 

before: OK (123 tests, 144 assertions)
after: Tests: 443, Assertions: 678, Skipped: 1.

Testing instructions:
Code review + run phpunit

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -45,6 +45,11 @@
             "Joomla\\CMS\\": "libraries/src/"
         }
     },
+    "autoload-dev": {
+        "psr-4": {
+            "Tests\\": "tests"
```
Comment (mbabker) [inline composer.json]: Why a different root namespace?  The general convention across Joomla's code (and a bulk of the PHP ecosystem) is to use your root namespace then put all tests in `\Tests` namespace below that, i.e. `Joomla\CMS\Tests`.

Comment (dneukirchen) [reply composer.json]: Fine for me... but then we will end up with Joomla\CMS\Tests\Unit\Libraries\Cms\Toolbar\Button...
I would prefer Joomla\Tests as root namespace.

**Comment by dgrammatiko** [comment]
@dneukirchen are you gonna bring the javascript tests as well? Do you want me to do aPR for that?
Is that also considered at this point?

**Comment by dneukirchen** [comment]
@dgrammatiko yes, the plan is to move all test suites back to repo. would be nice if you can do the js part. when you merge js tests back, it would be nice to update karma because of an sec issue in one of its dev dependencies. tried to reach you on glip to talk about it. we are using karma 0.13, current version is 4. karma 2+ should fix it.

**Comment by dgrammatiko** [comment]
@dneukirchen ok, I will try to do a PR this weekend

**Comment by TobsBobs** [comment]
What is the difference between this pull request and #23962?

**Comment by dneukirchen** [comment]
* it includes more unit tests (re-enabled some of the old unit tests which did not run for j4)
* tests are namespaced and psr-4 autoloaded
* tests use namespaced classes
* useless tests and integration tests in unit test suite are removed
* refactored some of the tests

**Comment by Hackwar** [comment]
While we are at it: Appveyor has a bunch of errors in its output. I doubt that it should be like that...

File: tests/Unit/Libraries/Cms/Access/RuleTest.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+/**
+ * Test class for \Joomla\CMS\Access\Rule.
+ *
+ * @package  Joomla.Platform
+ *
+ * @since    11.1
+ */
+class RuleTest extends UnitTestCase
+{
+	/**
```
Comment (brianteeman) [inline tests/Unit/Libraries/Cms/Access/RuleTest.php:24]: looks like there is an extra leading tab here

File: tests/Unit/Libraries/Cms/Access/RulesTest.php [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+		];
+
+		$rules = new Rules(  (object) $input);
+
+		$editRule = $rules->getData()['edit'];
+		$this->assertInstanceOf(Rule::class,  $editRule);
+		$this->assertEquals($ruleIdentities, $editRule->getData());
+	}
+
+	/**
```
Comment (brianteeman) [inline tests/Unit/Libraries/Cms/Access/RulesTest.php:88]: looks like there is an extra leading tab here

**Comment by rdeutz** [comment]
looks good to me

File: tests/Unit/Libraries/Cms/Access/RuleTest.php (hunk 2 above)
Comment (infograf768) [reply tests/Unit/Libraries/Cms/Access/RuleTest.php:24]: rather missing tabs below

File: tests/Unit/Libraries/Cms/Access/RulesTest.php (hunk 3 above)
Comment (infograf768) [reply tests/Unit/Libraries/Cms/Access/RulesTest.php:88]: same comment as for the other one

**Comment by Hackwar** [comment]
> What is the difference between this pull request and #23962?

I'll jsut leave this here: https://github.com/joomla/test-unit/pull/9

File: phpunit.xml.dist [hunk 4]
```diff
@@ -1,51 +1,12 @@
...
-			<directory>tests/unit/suites/libraries/cms</directory>
-		</testsuite>
-		<testsuite name="libraries-platform">
-			<directory>tests/unit/suites/libraries/joomla</directory>
-		</testsuite>
-		<testsuite name="administrator">
-			<directory>tests/unit/suites/administrator</directory>
-		</testsuite>
-		<testsuite name="FinderIndexer">
-			<directory>tests/unit/suites/finderIndexer</directory>
```
Comment (Hackwar) [inline phpunit.xml.dist:27]: correct me if I'm wrong, but unit and integration tests will both be run by phpunit, so we have to distinguish the tests by testsuite. I also will provide tests again for the Finder Indexer, so we would have to have several runs of the test suite(s). Afaik we can only have this one phpunit.xml and also just the one bootstrap.php, so we should keep that in mind for the different tests.

File: tests/Unit/Libraries/Cms/Environment/BrowserTest.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+			],
+			'Firefox 12 Android Tablet' => [
+				'Mozilla/5.0 (Android; Tablet; rv:12.0) Gecko/12.0 Firefox/12.0',
+				'firefox',
+				'unix',
+				'12',
+				true,
+			],
+			'Firefox 6 Windows' => [
+				'Mozilla/5.0 (Windows NT 5.0; WOW64; rv:6.0) Gecko/20100101 Firefox/6.0',
```
Comment (Hackwar) [inline tests/Unit/Libraries/Cms/Environment/BrowserTest.php:130]: So my current FF version is 65.0.1... Could we update this to contain something newer than 2011?

File: phpunit.xml.dist (hunk 4 above)
Comment (dneukirchen) [reply phpunit.xml.dist:27]: both run by phpunit, but we can have different config and bootstrap files per test suite. 
and i would recommend this. we dont need to start the app for unit tests... 
tbh: as long as we dont have a working integration test setup, i dont care that much about the suite config in phpunit.xml. atm im not sure if we should move the integration tests to codeception or keep it in plain old phpunit...

its possible to have

`phpunit --configuration unit.phpunit.xml tests/Unit`
`phpunit --configuration integration.phpunit.xml tests/Integration`

File: tests/Unit/Libraries/Cms/Environment/BrowserTest.php (hunk 5 above)
Comment (dneukirchen) [reply tests/Unit/Libraries/Cms/Environment/BrowserTest.php:130]: can you update the datasets and open pr against the base branch? would be great!

File: composer.json (hunk 1 above)
Comment (dneukirchen) [reply composer.json]: changed it to Joomla\Tests

**Comment by dneukirchen** [comment]
Rtc?

**Comment by dgrammatiko** [comment]
@dneukirchen cross referencing 
https://github.com/joomla/joomla-cms/pull/24044

**Comment by dneukirchen** [comment]
fixed merge conflict.

**Comment by dneukirchen** [comment]
ping @rdeutz @wilsonge @HLeithner can we get this merged?

**Review by rdeutz — APPROVED** [review]
Looks good to me

**Comment by dgrammatiko** [comment]
@rdeutz also #24044

**Comment by rdeutz** [comment]
@dgrammatiko I am on it

**Comment by infograf768** [comment]
Hmm.
`tests/Unit/Libraries/Cms/Mircodata/MicrodataTest.php`

`Mircodata` ??

Also it looks like these new files have not been added as indexed to the branch.

**Pull request merged**